src/app.tsx

```diff
@@ -0,0 +1,438 @@
+  // also sets the id for a newest spawned bug
+  const numSpawnedBugs = useRef(0);
+
+  // component did mount
+  useEffect(() => {
+    // create a new bug and give it a unique index
+
+    const spawnNewBug = () => {
+      // if (n > UPPER_COUNT) return
+      if (bugIds.length >= UPPER_BUG_COUNT) {
+        return;
+      } else {
+        numSpawnedBugs.current += 1;
+        setBugIds((prevState) => {
+          return [...prevState, numSpawnedBugs.current];
+        });
+      }
+    };
+
+    intervalID.current = window.setInterval(spawnNewBug, 1000);
+
+    return () => {
+      clearInterval(intervalID.current);
+    };
+  }, [bugIds, numSpawnedBugs]);
+
+  /**
+   * toggles the render state of modal component
+   */
+  const toggleModal = () => {
+    setShowModal((prevState) => !prevState);
+  };
+
+  /**
+   * Delivers info about current clicked bug to Modal and removes the bug from screen
+   *
+   * @param bugId - the id of the bug so that it can be deleted from bugs array
+   */
+  const handleBug = (bugType: string, bugId: number) => {
+    toggleModal();
+
+    setCurrentBugType(bugType);
+
+    if (BUGS[bugType]) {
+      if (!BUGS[bugType].isFriendly) {
+        if (!hasWon) {
+          if (!correctBugs.has(bugType)) {
+            // does not have bugId yet
+            if (correctBugs.add(bugType).size === TARGET_CORRECT_GUESSES) {
+              // check for win on next guess
+              // if win, don't reset yet so that progress bar shows full right after winning
+              setHasWon(true);
+              setCorrectBugs((prevState) => {
+                return new Set([...prevState, bugType]);
+              });
+            } else {
+              // if next guess not win, just add to list
+              setHasWon(false);
+              setCorrectBugs((prevState) => {
+                return new Set([...prevState, bugType]);
+              });
+            }
+          } else {
+            // already has bugId
+            // do nothing
+          }
+        } else {
+          // reset if u click bad bug after winning game
+          setHasWon(false);
+          setCorrectBugs(new Set());
+          setCorrectBugs((prevState) => {
+            return new Set([...prevState, bugType]);
+          });
+        }
+      } else {
+        // if bug is not friendly
+        if (hasWon) {
+          // reset if u click good bug after winning game
+          setHasWon(false);
+          setCorrectBugs(new Set());
+        } else {
+          // this triggers when correctBugs set is either empty or not-yet-full
+          // two possibilities
+          // resets all progress if u click on a good bug
+          setHasWon(false);
+          setCorrectBugs(new Set());
+
+          // or keep the current progress (keep the set)
+          //   setHasWon(false);
+        }
+      }
+    } else {
+      throw new Error("Whoops! The given BugId does not exist in BUGS object.");
+    }
```

Comment:
After refactor, long `if-else` statement above should become:

```
// handle error case early here

if (!BUGS[bugType].isFriendly) {
    handleCorrectBugClick();
} else {
    handleIncorrectBugClick();
}

```

Note: Pass required parameters to both `handleCorrectBugClick(...)`  function.